### PR TITLE
[utils] Backport traverse_obj (etc) from yt-dlp

### PR DESCRIPTION
**Pull request by lebdron**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Backport `traverse_obj` from yt-dlp required for https://github.com/ytdl-org/youtube-dl/pull/31097

### CODE REVIEW

**Comment by dirkf** [comment]
Please use my full port in [PR #30713](https://github.com/ytdl-org/youtube-dl/blob/d799c405010e50f75e253ace218de9c4be8744a9/youtube_dl/extractor/ceskatelevize.py#L26) as a basis. Feel free to define `variadic()` in this PR too, instead of locally within `traverse_obj()`, and to define `IDENTITY()`, `get_first()` and join_nonempty()` as well.

The `from_iterable()` function in my code is needed because the library function used in yt-dlp isn't always available.

The `Nonlocal` thing is to simulate the `nonlocal` syntax for older Pythons. As you know, assigning to a previously undefined variable creates a variable binding in the current scope, unless the variable has been defined `global`. `nonlocal` tells Python to look for the variable on the call stack too, so that a new binding is not created when assigning to the variable. By making the variable that is to be nonlocal an instance value of an object, an assignment to the variable doesn't create a new binding, but updates the existing value.

`listish()` defines an expression used in several places so that the list of listish things can be updated with one change.

As you've done, `...` has to be replaced with `Ellipsis` as the `...` syntax is only in recent Pythons.

Finally, any `str` should be `compat_str`.

Even more finally, should we ask @pukkandan to write tests for it?

The code in PR #30713 was correct enough for that extractor, but as was commented extractors only use a subset of ``traverse_obj()` functionality.

**Comment by lebdron** [comment]
@dirkf Thank you, done. I looked at your commit, ported some features which improve readability, like `Nonlocal`. I also used the new code from https://github.com/yt-dlp/yt-dlp/commit/1797b073ed9989d57a84a3e05106285c54085686 and https://github.com/yt-dlp/yt-dlp/commit/e6f868a63c15f576152733a1508f474b5e5bd1ef .

I would suggest to merge this PR as is, without extensive test coverage, as it will be useful to port Panopto sooner, and it does not break any existing code. Tests can be added later in a separate pull request.

**Comment by pukkandan** [comment]
> Even more finally, should we ask @pukkandan to write tests for it?

I honestly don't have too much incentive to write tests for it. Because of the extensive use of `traverse_obj` in "output template" in yt-dlp, [test_prepare_outtmpl_and_filename](https://github.com/yt-dlp/yt-dlp/blob/07a1250e0e90515ff8142161536f9dafa6eaba1b/test/test_YoutubeDL.py#L672) kinda acts like a decent test suite for `traverse_obj`. But if anyone wants to contribute tests, it is welcome.

**Comment by pukkandan** [comment]
Complete rewrite and tests: https://github.com/yt-dlp/yt-dlp/pull/5024

**Comment by dirkf** [comment]
Good news: there are tests
Bad news: the back-ported code should match the new code

**Review by dirkf — COMMENTED** [review]
Recast with yt-dlp's new traverse_obj() implementation and **tests**.

File: test/test_utils.py [hunk 1]
```diff
@@ -1475,6 +1487,281 @@ def test_clean_podcast_url(self):
...
+                         msg='failing str key on a `re.Match` should return `default`')
+        self.assertEqual(traverse_obj(mobj, 8), None,
+                         msg='failing int key on a `re.Match` should return `default`')
+
+    def test_get_first(self):
+        self.assertEqual(get_first([{'a': 'b'}], 'a'), 'b')
+
+    def test_join_nonempty(self):
+        self.assertEqual(join_nonempty('a', 'b'), 'a-b')
+        self.assertEqual(join_nonempty('a', 'b', from_dict={'a': 'c', 'b': 'd'}), 'c-d')
```
Comment (dirkf) [inline test/test_utils.py:1765]:
```suggestion


```

File: youtube_dl/utils.py [hunk 2]
```diff
@@ -5801,3 +5923,218 @@ def clean_podcast_url(url):
...
+    get_all = kwargs.get('get_all', True)
+    casesense = kwargs.get('casesense', True)
+    _is_user_input = kwargs.get('_is_user_input', False)
+    _traverse_string = kwargs.get('_traverse_string', False)
+
+    # instant compat
+    str = compat_str
+
+    is_sequence = lambda x: isinstance(x, compat_collections_abc.Sequence) and not isinstance(x, (str, bytes))
+    # stand-in until casefold.py is added
```
Comment (dirkf) [inline youtube_dl/utils.py:5993]:
```suggestion
    # stand-in until compat_re_Match is added
    compat_re_Match = type(re.match('a', 'a'))
    # stand-in until casefold.py is added
```

File: youtube_dl/utils.py [hunk 3]
```diff
@@ -5801,3 +5923,220 @@ def clean_podcast_url(url):
...
+
+            if not is_sequence(obj):
+                if not _traverse_string:
+                    return
+                obj = str(obj)
+
+            try:
+                yield obj[key]
+            except IndexError:
+                pass
```
Comment (dirkf) [inline youtube_dl/utils.py:6091]:
```suggestion
            if not is_sequence(obj):
                if not _traverse_string:
                    return
                obj = str(obj)

            try:
                yield obj[key]
            except IndexError:
                pass
```

File: youtube_dl/utils.py [hunk 4]
```diff
@@ -5801,3 +5923,220 @@ def clean_podcast_url(url):
...
+
```
Comment (dirkf) [inline youtube_dl/utils.py:6092]:
```suggestion
            # force change for Linter
            if not is_sequence(obj):
                if not _traverse_string:
                    return
                obj = str(obj)

            try:
                yield obj[key]
            except IndexError:
                pass

```

File: test/test_utils.py [hunk 5]
```diff
@@ -1475,6 +1487,282 @@ def test_clean_podcast_url(self):
...
+                'KeY': 'value1',
+                0: {'KeY': 'value2'},
+            },
+        }
+        self.assertEqual(traverse_obj(_CASESENSE_DATA, 'key'), None,
+                         msg='dict keys should be case sensitive unless `casesense`')
+        self.assertEqual(traverse_obj(_CASESENSE_DATA, 'keY',
+                                      casesense=False), 'value0',
+                         msg='allow non matching key case if `casesense`')
+        self.assertEqual(traverse_obj(_CASESENSE_DATA, (0, ('keY',)),
```
Comment (dirkf) [inline test/test_utils.py:1696]:
```suggestion
                         msg='allow non matching key case if `casesense`')
        self.assertEqual(traverse_obj(_CASESENSE_DATA, '\uff4bey',  # FULLWIDTH LATIN SMALL LETTER K
                                      casesense=False), 'value3',
                         msg='allow non matching Unicode key case if `casesense`')
        self.assertEqual(traverse_obj(_CASESENSE_DATA, (0, ('keY',)),
```

Comment (dirkf) [inline test/test_utils.py:1696]:
```suggestion
        # Test casesense behavior
        _CASESENSE_DATA = {
            'KeY': 'value0',
            0: {
                'KeY': 'value1',
                0: {'KeY': 'value2'},
            },
             # FULLWIDTH LATIN CAPITAL LETTER K
            '\uff2bey': 'value3',
        }
        self.assertEqual(traverse_obj(_CASESENSE_DATA, 'key'), None,
                         msg='dict keys should be case sensitive unless `casesense`')
        self.assertEqual(traverse_obj(_CASESENSE_DATA, 'keY',
                                      casesense=False), 'value0',
                         msg='allow non matching key case if `casesense`')
        self.assertEqual(traverse_obj(_CASESENSE_DATA, '\uff4bey',  # FULLWIDTH LATIN SMALL LETTER K
                                      casesense=False), 'value3',
                         msg='allow non matching Unicode key case if `casesense`')
        self.assertEqual(traverse_obj(_CASESENSE_DATA, (0, ('keY',)),
```

File: test/test_utils.py [hunk 6]
```diff
@@ -79,6 +84,7 @@
     strip_or_none,
     subtitles_filename,
     timeconvert,
+    traverse_obj,
```
Comment (dirkf) [inline test/test_utils.py:87]:
```suggestion
    traverse_obj,
    try_call,
```

File: test/test_utils.py [hunk 7]
```diff
@@ -1475,6 +1487,287 @@ def test_clean_podcast_url(self):
...
+        ll = LazyList(range(10), reverse=True)
+        test(ll, -1, 0, range(1))
+        test(ll, 3, 6, range(10))
+
+        ll = LazyList(itertools.count())
+        test(ll, 10, 10, range(11))
+        ll = reversed(ll)
+        test(ll, -15, 14, range(15))
+
+    def test_traverse_obj(self):
```
Comment (dirkf) [inline test/test_utils.py:1535]:
```suggestion
    def test_try_call(self):
        def total(*x, **kwargs):
            return sum(x) + sum(kwargs.values())

        self.assertEqual(try_call(None), None,
                         msg='not a fn should give None')
        self.assertEqual(try_call(lambda : 1), 1,
                         msg='int fn with no expected_type should give int')
        self.assertEqual(try_call(lambda : 1, expected_type=int), 1,
                         msg='int fn with expected_type int should give int')
        self.assertEqual(try_call(lambda : 1, expected_type=dict), None,
                         msg='int fn with wrong expected_type should give None')
        self.assertEqual(try_call(total, args=(0, 1, 0, ), expected_type=int), 1,
                         msg='fn should accept arglist')
        self.assertEqual(try_call(total, kwargs={'a': 0, 'b': 1, 'c': 0}, expected_type=int), 1,
                         msg='fn should accept kwargs')
        self.assertEqual(try_call(lambda : 1, expected_type=dict), None,
                         msg='int fn with no expected_type should give None')
        self.assertEqual(try_call(lambda x: {}, total, args=(42, ), expected_type=int), 42,
                         msg='expect first int result with expected_type int')

    def test_variadic(self):
        self.assertEqual(variadic(None), (None, ))
        self.assertEqual(variadic('spam'), ('spam', ))
        self.assertEqual(variadic('spam', allowed_types=dict), 'spam')

    def test_traverse_obj(self):
```

File: test/test_utils.py [hunk 8]
```diff
@@ -1475,6 +1487,287 @@ def test_clean_podcast_url(self):
...
+                         msg='tuple as key should be treated as branches')
+        self.assertEqual(traverse_obj(_TEST_DATA, ('urls', [3, 0], 'url')), ['https://www.example.com/0'],
+                         msg='list as key should be treated as branches')
+        self.assertEqual(traverse_obj(_TEST_DATA, ('urls', ((1, 'fail'), (0, 'url')))), ['https://www.example.com/0'],
+                         msg='double nesting in path should be treated as paths')
+        self.assertEqual(traverse_obj(['0', [1, 2]], [(0, 1), 0]), [1],
+                         msg='do not fail early on branching')
+        self.assertCountEqual(traverse_obj(_TEST_DATA, ('urls', ((1, ('fail', 'url')), (0, 'url')))),
+                              ['https://www.example.com/0', 'https://www.example.com/1'],
+                              msg='tripple nesting in path should be treated as branches')
```
Comment (dirkf) [inline test/test_utils.py:1608]:
```suggestion
                              msg='triple nesting in path should be treated as branches')
```

File: test/test_utils.py [hunk 9]
```diff
@@ -1475,6 +1487,287 @@ def test_clean_podcast_url(self):
...
+                         msg='str key on a `re.Match` should respect casesense')
+        self.assertEqual(traverse_obj(mobj, 'fail'), None,
+                         msg='failing str key on a `re.Match` should return `default`')
+        self.assertEqual(traverse_obj(mobj, 'gRoUpS', casesense=False), None,
+                         msg='failing str key on a `re.Match` should return `default`')
+        self.assertEqual(traverse_obj(mobj, 8), None,
+                         msg='failing int key on a `re.Match` should return `default`')
+
+    def test_get_first(self):
+        self.assertEqual(get_first([{'a': 'b'}], 'a'), 'b')
```
Comment (dirkf) [inline test/test_utils.py:1765]:
```suggestion
        self.assertEqual(get_first([{'a': None}, {'a': 'spam'}], 'a'), 'spam')
```

File: test/test_utils.py [hunk 10]
```diff
@@ -1475,6 +1487,287 @@ def test_clean_podcast_url(self):
...
+
+    def test_join_nonempty(self):
+        self.assertEqual(join_nonempty('a', 'b'), 'a-b')
+        self.assertEqual(join_nonempty('a', 'b', from_dict={'a': 'c', 'b': 'd'}), 'c-d')
```
Comment (dirkf) [inline test/test_utils.py:1769]:
```suggestion
        self.assertEqual(join_nonempty(
                'a', 'b', 'c', 'd',
                from_dict={'a': 'c', 'c': [], 'b': 'd', 'd': None}), 'c-d')
```

File: test/test_utils.py [hunk 11]
```diff
@@ -1475,6 +1488,287 @@ def test_clean_podcast_url(self):
...
+                         msg='do not overflatten if not `get_all`')
+
+        # Test casesense behavior
+        _CASESENSE_DATA = {
+            'KeY': 'value0',
+            0: {
+                'KeY': 'value1',
+                0: {'KeY': 'value2'},
+            },
+             # FULLWIDTH LATIN CAPITAL LETTER K
```
Comment (dirkf) [inline test/test_utils.py:1691]:
```suggestion
            # FULLWIDTH LATIN CAPITAL LETTER K
```

File: test/test_utils.py [hunk 12]
```diff
@@ -92,6 +98,7 @@
     urlencode_postdata,
     urshift,
     update_url_query,
+    # variadic,
```
Comment (dirkf) [inline test/test_utils.py:101]:
```suggestion
    variadic,
```

File: test/test_utils.py [hunk 13]
```diff
@@ -1475,6 +1489,315 @@ def test_clean_podcast_url(self):
...
+                         msg='failing int key on a `re.Match` should return `default`')
+
+    def test_get_first(self):
+        self.assertEqual(get_first([{'a': None}, {'a': 'spam'}], 'a'), 'spam')
+
+    def test_join_nonempty(self):
+        self.assertEqual(join_nonempty('a', 'b'), 'a-b')
+        self.assertEqual(join_nonempty(
+                'a', 'b', 'c', 'd',
+                from_dict={'a': 'c', 'c': [], 'b': 'd', 'd': None}), 'c-d')
```
Comment (dirkf) [inline test/test_utils.py:1799]:
```suggestion
              'a', 'b', 'c', 'd',
              from_dict={'a': 'c', 'c': [], 'b': 'd', 'd': None}), 'c-d')
```

**Pull request merged**